src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs

```diff
@@ -71,15 +73,32 @@ public TextDocumentState(DocumentInfo info, LoadTextOptions loadTextOptions, Hos
         public IReadOnlyList<string> Folders => Attributes.Folders;
         public string Name => Attributes.Name;
 
-        private static ITextAndVersionSource CreateStrongText(TextAndVersion text)
-            => new ConstantTextAndVersionSource(text);
+        private static readonly ConditionalWeakTable<SourceText, ITextAndVersionSource> s_textToStrongTextAndVersionSource = new();
+        private static readonly ConditionalWeakTable<TextLoader, ITextAndVersionSource> s_textLoaderToStrongTextAndVersionSource = new();
+        private static readonly ConditionalWeakTable<SourceText, ITextAndVersionSource> s_textToRecoverableTextAndVersionSource = new();
+        private static readonly ConditionalWeakTable<TextLoader, ITextAndVersionSource> s_textLoaderToRecoverableTextAndVersionSource = new();
 
-        private static ITextAndVersionSource CreateStrongText(TextLoader loader)
-            => new LoadableTextAndVersionSource(loader, cacheResult: true);
+        private static ITextAndVersionSource GetOrCreateStrongText(TextAndVersion textAndVersion)
+        {
+            // Note: it is intentional that we discard the version of textAndVersion when looking to see if we can find
+            // an existing cached ITextAndVersionSource.  The version effectively corresponds to what version the *text*
+            // is at, so if we have two cases pointing at the exact same SourceText, then it is appropriate that the
+            // versions stay the same.
```

Comment:
i'm mostly convinced this is safe/sane.  *Today* the version we pass in here anyways comes from a document saying "GetNewerVersion".  That newer version isn't held onto it at all, and is only kept here.  The change in version just seems to be a way to update the version stamp.  With this change it means the first doc in a linked-doc series will 'win' on the version, and then the other links will haev the same version.  That seems sensible to me since they all should correspond to the same entity effectively.